util/src/main/java/org/datacommons/util/DcidGenerator.java

```diff
@@ -0,0 +1,369 @@
+  // Max number of bits of a 32-bit encoding
+  private static final int MAX_NUM_BITS = 5;
+
+  // Set of place types that use random short-ID. Typically though their IDs are derived from
+  // external IDs (like geoId/06).
+  private static final Set<String> TYPES_USING_SHORT_ID =
+      Set.of(
+          "City",
+          "County",
+          "Province",
+          "State",
+          "Country",
+          "Continent",
+          "CollegeOrUniversity",
+          "Source",
+          "Curator",
+          "Provenance",
+          "AdministrativeArea",
+          "AdministrativeArea1",
+          "AdministrativeArea2",
+          "AdministrativeArea3",
+          "AdministrativeArea4",
+          "AdministrativeArea5",
+          "Place",
+          "Town",
+          "Village",
+          "Neighborhood");
+
+  private static final Set<String> FIXED_STAT_VAR_PROPS =
+      Set.of(
+          Vocabulary.TYPE_OF,
+          Vocabulary.POPULATION_TYPE,
+          Vocabulary.MEASURED_PROP,
+          Vocabulary.STAT_TYPE,
+          Vocabulary.MEASUREMENT_QUALIFIER,
+          Vocabulary.MEASUREMENT_DENOMINATOR);
+
+  static final int LAST_REQUIRED_STAT_VAR_OBS_PROP_INDEX = 3; // until value
+  private static final List<String> ORDERED_STAT_VAR_OBS_KEY_PROPS =
+      List.of(
+          Vocabulary.OBSERVATION_ABOUT,
+          Vocabulary.VARIABLE_MEASURED,
+          Vocabulary.OBSERVATION_DATE,
+          Vocabulary.GENERIC_VALUE,
+          Vocabulary.OBSERVATION_PERIOD,
+          Vocabulary.UNIT,
+          Vocabulary.MEASUREMENT_METHOD,
+          Vocabulary.SCALING_FACTOR);
+
+  static final int LAST_REQUIRED_LEGACY_OBS_PROP_INDEX = 2; // until measuredProp.
+  private static final List<String> ORDERED_LEGACY_OBS_KEY_PROPS =
+      List.of(
+          Vocabulary.OBSERVED_NODE,
+          Vocabulary.OBSERVATION_DATE,
+          Vocabulary.MEASURED_PROP,
+          Vocabulary.OBSERVATION_PERIOD,
+          Vocabulary.UNIT,
+          Vocabulary.MEASURED_VALUE,
+          Vocabulary.MIN_VALUE,
+          Vocabulary.MAX_VALUE,
+          Vocabulary.SUM_VALUE,
+          Vocabulary.MEAN_VALUE,
+          Vocabulary.MEDIAN_VALUE,
+          Vocabulary.MARGIN_OF_ERROR,
+          Vocabulary.MEASUREMENT_METHOD,
+          Vocabulary.MEASUREMENT_RESULT,
+          Vocabulary.STD_ERROR,
+          Vocabulary.SAMPLE_SIZE,
+          Vocabulary.PERCENTILE_10,
+          Vocabulary.PERCENTILE_25,
+          Vocabulary.PERCENTILE_75,
+          Vocabulary.PERCENTILE_90,
+          Vocabulary.MEAN_STD_ERROR,
+          Vocabulary.GROWTH_RATE,
+          Vocabulary.STD_DEVIATION_VALUE,
+          Vocabulary.MEASUREMENT_DENOMINATOR,
+          Vocabulary.MEASUREMENT_QUALIFIER,
+          Vocabulary.SCALING_FACTOR);
+
+  private static final LogWrapper dummyLogCtx =
+      new LogWrapper(Debug.Log.newBuilder(), Path.of("/tmp/report.html"));
+
+  public static class Result {
+    public String dcid = new String();
```

Comment:
maybe also add a comment here that the result could be an empty dcid if there are errors

Reply:
Done.